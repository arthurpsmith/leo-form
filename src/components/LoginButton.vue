<template>
  <span>
    <v-btn
      v-if="!isLoggedIn"
      color="primary"
      class="ml-4"
      @click="attemptLogin"
    >{{ buttonText }}</v-btn>
    <v-menu
      v-if="isLoggedIn"
      class="ml-4"
    >
      <template v-slot:activator="{ on, attrs }">
        <v-btn
          outlined
          v-bind="attrs"
          v-on="on"
          class="ml-4"
        >
          {{ userName }}
          <v-icon>mdi-menu-down</v-icon>
        </v-btn>
      </template>
      <v-list>
        <v-list-item @click="authMachine.send('LOG_OUT')">
          <v-list-item-title>{{ buttonText }}</v-list-item-title>
        </v-list-item>
      </v-list>
    </v-menu>
</span>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api'
import { useAuthService } from '@/statemachines/auth.machine'
import { StateValue } from 'xstate'

export default defineComponent({
  name: 'LoginButton',
  setup () {
    const authMachine = useAuthService()
    const attemptLogin = () => {
      authMachine.send({
        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
        // @ts-ignore
        type: 'ATTEMPT_LOG_IN',
        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
        // @ts-ignore
        userDetails: {
          /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
          // @ts-ignore
          username: 'someone@crossref.org'
        }
      })
    }
    return {
      authMachine, attemptLogin
    }
  },
  computed: {
    authState (): StateValue {
      return this.authMachine.state.value.value
    },
    userName (): string {
      return (typeof this.authMachine.state.value?.context?.userDetails !== 'undefined') ? this.authMachine.state.value.context.userDetails.username : ''
    },
    buttonText (): string {
      switch (this.authState) {
        case 'loggedIn':
          return 'Log out'
        default:
          return 'Log in'
      }
    },
    isLoggedIn (): boolean {
      return this.authState === 'loggedIn'
    }
  }
})
</script>
