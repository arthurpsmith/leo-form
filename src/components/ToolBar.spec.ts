import { createLocalVue, mount } from '@vue/test-utils'
import ToolBar from '@/components/ToolBar.vue'
import Vuetify from 'vuetify'
import CompositionApi from '@vue/composition-api'
import { authSymbol, getAuthService } from '@/statemachines/auth.machine'
import { useActor } from 'xstate-vue2'
import LoginButton from '@/components/LoginButton.vue'
import { getSearchService, searchSymbol } from '@/statemachines/search.machine'

describe('ToolBar.vue', () => {
  // DO NOT use Vuetify on the localInstance
  // This is bootstrapped in the jest setup
  // file located in ./tests/setup.ts
  //
  // localVue.use(Vuetify)

  const localVue = createLocalVue()
  localVue.use(CompositionApi)
  let vuetify: Vuetify
  let authServiceToProvide: ReturnType<typeof getAuthService>
  let provideWithExportedSymbol: Record<string, unknown>
  let searchServiceToProvide: ReturnType<typeof getSearchService>
  let authService
  let searchService

  function mountFunction (options = {}) {
    return mount(ToolBar, {
      localVue,
      vuetify,
      provide: provideWithExportedSymbol,
      ...options
    })
  }

  beforeEach(() => {
    vuetify = new Vuetify()
    authServiceToProvide = getAuthService()
    searchServiceToProvide = getSearchService()
    provideWithExportedSymbol = {
      [authSymbol as symbol]: authServiceToProvide,
      [searchSymbol as symbol]: searchServiceToProvide
    }
    authService = useActor(authServiceToProvide)
  })

  it('should have a custom title and match snapshot', () => {
    const titleText = 'Foobar title'
    const wrapper = mountFunction({
      propsData: {
        title: titleText
      }
    })

    // With jest we can create snapshot files of the HTML output
    expect(wrapper.html()).toMatchSnapshot()

    // We could also verify this differently
    // by checking the text content
    const title = wrapper.find('.v-toolbar__title')

    expect(title.text()).toBe(titleText)
  })
})
