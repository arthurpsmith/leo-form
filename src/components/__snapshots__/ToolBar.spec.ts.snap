// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolBar.vue should have a custom title and match snapshot 1`] = `
<header class="v-sheet theme--dark v-toolbar v-toolbar--dense secondary" style="height: 48px;">
  <div class="v-toolbar__content" style="height: 48px;"><button type="button" class="v-app-bar__nav-icon v-btn v-btn--icon v-btn--round theme--dark v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu theme--dark"></i></span></button>
    <div class="v-toolbar__title">Foobar title</div>
    <div class="spacer"></div>
    <div class="v-input v-input--hide-details v-input--dense theme--dark v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--filled v-text-field--enclosed v-text-field--outlined v-text-field--placeholder expanding-search closed">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-input__prepend-inner">
            <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--dark"></i></div>
          </div>
          <fieldset aria-hidden="true">
            <legend><span>â€‹</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input data-cy="search-box__input" id="input-8" placeholder="Search" type="text"></div>
          <div class="v-input__append-inner">
            <div></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
`;
